### PR TITLE
Avoid custom hidden fields, avoid break replaced models

### LINKED ISSUE #1816: [Bug? 4.3.0] Checking Permission via can() return PermissionDoesNotExists
Hi. First, thanks for this great packages. I've been using this packages since version 1.

I have user with Role 'System Administrator', and this roles has permission named 'dashboard', 'logs', etc. When I'm checking the user's permission via `can`, it throws `Spatie\Permission\Exceptions\PermissionDoesNotExist`.

- spatie/laravel-permission package version: 4.3.0
- illuminate/framework package: v8.55.0

PHP version: 8.0.7
Database version: MySQL 8.0.25

This is not happening prior to version 4.3.0. We have traced this issue, and found that `getPermissions` on `Spatie\Permission\PermissionRegistrar` is the culprit here.

This is how we resolve this problem:
```php
        $permissions = $this->permissions->$method(static function ($permission) use ($params) {
            foreach ($params as $attr => $value) {
                if ($permission->getAttribute($attr) == $value) {
                    return true;
                }
            }

            return false;
        });
```

I think the return value are swapped? If I swap the 'true' and 'false' value like above, it behaves normally, and I only get `PermissionDoesNotExists` exception only when I'm supplying it with wrong / invalid permission.

Thanks

### LINKED ISSUE #1820: Failing tests with RefreshDatabase as "permission already exists"
**Describe the bug**
I would like to perform a test with `RefreshDatabase`; if I do so, I am getting this error:

```
 A `index dashboard` permission already exists for guard `web`.

  at vendor/spatie/laravel-permission/src/Exceptions/PermissionAlreadyExists.php:11
      7▕ class PermissionAlreadyExists extends InvalidArgumentException
      8▕ {
      9▕     public static function create(string $permissionName, string $guardName)
     10▕     {
  ➜  11▕         return new static("A `{$permissionName}` permission already exists for guard `{$guardName}`.");
     12▕     }
     13▕ }
     14▕ 

      +1 vendor frames 
  2   database/seeders/RolesAndPermissionsSeeder.php:22
      Spatie\Permission\Models\Permission::create()

      +7 vendor frames 
  10  database/seeders/DatabaseSeeder.php:16
      Illuminate\Database\Seeder::call()
```

This is the `TestCase.php`

```
<?php

namespace Tests;

use App\Models\User;
use Illuminate\Foundation\Testing\TestCase as BaseTestCase;
use Illuminate\Support\Facades\Artisan;

/**
 * Class TestCase
 *
 * @package Tests
 */
abstract class TestCase extends BaseTestCase
{
    use CreatesApplication;

    /**
     * @var
     */
    protected $admin;

    /**
     * @var
     */
    protected $user_one;

    /**
     * @var
     */
    protected $user_two;

    /**
     * @var
     */
    protected $user_admin;

    /**
     * Set up the tests
     */
    public function setUp(): void
    {
        parent::setUp();
        Artisan::call('migrate', ['-vvv' => true]);
        Artisan::call('passport:install', ['-vvv' => true, '--uuids' => true, '--no-interaction' => true]);
        $this->seed();
        $this->user_admin = User::where('email', 'test@example.org')->first();
    }
}
```

This is the failing test:

```
<?php

namespace Tests\Feature;

use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

/**
 *
 */
class PrivacyTest extends TestCase
{
    use RefreshDatabase;

    /**
     * @var bool
     */
    public $mockConsoleOutput = false;

    /**
     * Test if the user can log in via HTTP API.
     *
     * @return void
     */
    public function test_user_can_view_own_user_data()
    {
        $user = $this->user_one;

        $this->json('GET', "/api/user/{$user->id}")->assertStatus(200)
             ->assertJsonStructure([
                 'token_type',
                 'expires_at',
                 'access_token',
             ]);
    }
}
```

This is the `DatabaseSeeder.php`

```
<?php

namespace Database\Seeders;

use Illuminate\Database\Seeder;

class DatabaseSeeder extends Seeder
{
    /**
     * Seed the application's database.
     *
     * @return void
     */
    public function run()
    {
        $this->call(RolesAndPermissionsSeeder::class);
        $this->call(UsersSeeder::class);
        $this->call(CompanySeeder::class);
    }
}
```

And this is the `RolesAndPermissionsSeeder.php`

```
<?php

namespace Database\Seeders;

use Illuminate\Database\Seeder;
use Spatie\Permission\Models\Permission;
use Spatie\Permission\Models\Role;
use Spatie\Permission\PermissionRegistrar;

class RolesAndPermissionsSeeder extends Seeder
{
    public function run()
    {
        /*reset cached roles and permissions*/
        app(PermissionRegistrar::class)->forgetCachedPermissions();

        /*Create permissions*/

        /*Dashboard*/
        Permission::create(['name' => 'index dashboard']);

        /*Add admin role*/
        $admin = Role::create(['name' => 'admin']);

        /*Dashboard*/
        $admin->givePermissionTo('index dashboard');
    }
}

```
**Versions**
You can use `composer show` to get the version numbers of:
- spatie/laravel-permission package version: 4.0.0

PHP version: 8
Database version: MariaDB10

 **Additional context**
I am using Laravel sail with this `.env.testing`

```
APP_NAME=project
APP_ENV=local
APP_KEY=base64:...
APP_DEBUG=true
APP_URL=http://project.test
APP_PORT=8084

LOG_CHANNEL=stack
LOG_LEVEL=debug

DB_CONNECTION=mysql
DB_HOST=project_mariadb_test_1
DB_PORT=3306
DB_DATABASE=testdatabase
DB_USERNAME=testuser
DB_PASSWORD=testpassword

TEST_DB_DATABASE=testdatabase
TEST_DB_USERNAME=testuser
TEST_DB_PASSWORD=testpassword

BROADCAST_DRIVER=log
CACHE_DRIVER=file
QUEUE_CONNECTION=sync
SESSION_DRIVER=file
SESSION_LIFETIME=120

MEMCACHED_HOST=memcached

REDIS_HOST=redis
REDIS_PASSWORD=null
REDIS_PORT=6379

MAIL_MAILER=smtp
MAIL_HOST=mailhog
MAIL_PORT=1025
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS=null
MAIL_FROM_NAME="${APP_NAME}"

AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_DEFAULT_REGION=us-east-1
AWS_BUCKET=

PUSHER_APP_ID=
PUSHER_APP_KEY=
PUSHER_APP_SECRET=
PUSHER_APP_CLUSTER=mt1

MIX_PUSHER_APP_KEY="${PUSHER_APP_KEY}"
MIX_PUSHER_APP_CLUSTER="${PUSHER_APP_CLUSTER}"

MIX_APP_NAME="${APP_NAME}"
```

and this `docker-compose.yml`

```
# For more information: https://laravel.com/docs/sail
version: '3'
services:
  laravel.test:
    build:
      context: ./vendor/laravel/sail/runtimes/8.0
      dockerfile: Dockerfile
      args:
        WWWGROUP: '${WWWGROUP}'
    image: sail-8.0/app
    ports:
      - '${APP_PORT:-80}:80'
    environment:
      WWWUSER: '${WWWUSER}'
      LARAVEL_SAIL: 1
    volumes:
      - '.:/var/www/html'
    networks:
      - sail
    depends_on:
      - mariadb
  mariadb:
    image: 'mariadb:10'
    ports:
      - '${FORWARD_DB_PORT:-3306}:3306'
    environment:
      MYSQL_ROOT_PASSWORD: '${DB_PASSWORD}'
      MYSQL_DATABASE: '${DB_DATABASE}'
      MYSQL_USER: '${DB_USERNAME}'
      MYSQL_PASSWORD: '${DB_PASSWORD}'
      MYSQL_ALLOW_EMPTY_PASSWORD: 'yes'
    volumes:
      - 'sailmariadb:/var/lib/mysql'
    networks:
      - sail
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-p${DB_PASSWORD}"]
      retries: 3
      timeout: 5s
  mariadb_test:
    image: 'mariadb:10'
    environment:
      MYSQL_ROOT_PASSWORD: "${TEST_DB_PASSWORD}"
      MYSQL_DATABASE: "${TEST_DB_DATABASE}"
      MYSQL_USER: "${TEST_DB_USERNAME}"
      MYSQL_PASSWORD: "${TEST_DB_PASSWORD}"
    networks:
      - sail
    healthcheck:
      test: [ "CMD", "mysqladmin", "ping" ]
networks:
  sail:
    driver: bridge
volumes:
  sailmariadb:
    driver: local
```
**Environment (please complete the following information, because it helps us investigate better):**
 - OS: macOS

### PR DESCRIPTION
**Pull request by erikn69**
Closes #1816
Avoid breaking change from #1799

### CODE REVIEW

**Comment by drbyte** [comment]
@erikn69 Just to confirm, this is temporary code due to temporary BC until next major version, right?

**Comment by erikn69** [comment]
```php
if (! method_exists($this->getPermissionClass(), 'getModelFromArray')) {
    return $permissions;
}
```
@drbyte the condition is temporary code due to temporary BC, if custom replaced models has method ´getModelFromArray´, they can use smaller cache, if not, they use the old collection cache without break

**Comment by drbyte** [comment]
> other changes fixes #1820

Did you mean #1816 ?

**Comment by erikn69** [comment]
Yes, sorry, i'm full of work here on, 
I confused the codes

File: tests/TestCase.php [hunk 1]
```diff
@@ -78,6 +78,8 @@ protected function getEnvironmentSetUp($app)
         // Set-up admin guard
         $app['config']->set('auth.guards.admin', ['driver' => 'session', 'provider' => 'admins']);
         $app['config']->set('auth.providers.admins', ['driver' => 'eloquent', 'model' => Admin::class]);
+        $app['config']->set('permission.models.permission', \Spatie\Permission\Test\Permission::class);
+        $app['config']->set('permission.models.role', \Spatie\Permission\Test\Role::class);
```
Comment (drbyte) [inline tests/TestCase.php:82]: Should we add a test that uses the base model directly as well ... to ensure there's no break in that case?

Comment (erikn69) [reply tests/TestCase.php:82]: I was thinking that, maybe is better remove that ?

Comment (drbyte) [reply tests/TestCase.php:82]: I guess it depends what we're testing! 😄 
By adding these models and forcing their use in tests, we've excluded directly testing the original model, which most applications will be using. :(

Comment (erikn69) [reply tests/TestCase.php:82]: So, better i let original model testing and make a new test only for test extended models

Comment (erikn69) [reply tests/TestCase.php]: @drbyte added new test `HasPermissionsWithCustomModelsTest.php` for only test custom models, better readability using collection and eloquent methods

File: tests/Permission.php [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace Spatie\Permission\Test;
+
+class Permission extends \Spatie\Permission\Models\Permission
```
Comment (incrize) [inline tests/Permission.php:5]: In my opinion, this is no different from testing with the original model and does not in any way check for possible backward incompatible changes.

Comment (erikn69) [reply tests/Permission.php:5]: [#1816](https://github.com/spatie/laravel-permission/issues/1816#issuecomment-903523159) It checks when some field is hidden on custom models(for now, can be used for avoid other break on custom models)
```php
protected $visible = [
      'id', 
      'name',
    ];
```

**Pull request merged**